keps/sig-network/4962-network-topology-standard/README.md

```diff
@@ -0,0 +1,867 @@
+
+In this KEP, we propose establishing a standardized representation of network topology within Kubernetes clusters.
+
+Such topology information could significantly enhance various Kubernetes components and features, including:
+- Pod affinity settings in deployments and pod specs.
+- Topology-aware scheduling in Kueue.
+- Development of Kubernetes-native scheduler plugins for network-topology-aware scheduling, such as:
+  - Topology-aware gang-scheduling plugin.
+  - Gang-scheduling auto-scaler.
+  - Dynamic resource allocation (DRA) scheduler plugin.
+
+### Goals
+
+- Introduce a standard way of representing network topology in Kubernetes clusters
+
+### Non-Goals
+
+- Implement a network-topology-aware gang-scheduling scheduler plugin
+- Define or implement network topology discovery mechanisms for CSPs or on-premises environments
+
+## Proposal
+
+We propose new node label type to capture network topology information:
+
+### Network Topology Label
+Format: `network.topology.kubernetes.io/<hierarchy-layer-type>: <name>`, where
+- `<hierarchy-layer-type>` defines the topology and characteristics of a network switch. The term `switch` may refer to a physical network device or a collection of closely connected devices with similar functionality.
+- `<name>` is a unique identifier for the hierarchy layer type.
+
+We propose to use the following four network hierarchy layer types:
+1. `accelerator`: Network interconnect for direct accelerator communication (e.g., Multi-node NVLink interconnect between NVIDIA GPUs)
+2. `block`: Rack-level switches connecting hosts in one or more racks as a block.
+3. `spine`: Spine-level switches connecting multiple blocks inside a datacenter.
+4. `datacenter`: Zonal switches connecting multiple datacenters inside an availability zone.
+
+These types will accommodate the majority of common network hierarchies across different CSP and on-prem environments.
+Having these labels available in Kubernetes clusters will help in designing cloud agnostic scheduling systems.
+The scheduler will prioritize hierarchy layers according to the order outlined above, providing a standardized
+approach for network-aware scheduling across a range of configurations.
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a data scientist running a data-intensive large-scale AI training job, I want to optimize the runtime
+by binding pods to nodes that are in close network proximity.
+This ensures better performance for my distributed workloads.
+
+Additionally, I do not want to modify the job specification when migrating the job across different Kubernetes environments.
+
+To achieve this, I can leverage the pod affinity feature of the default Kubernetes scheduler with topology keys:
```

Comment:
I think the using the beta DRA APIs would be a better way to represent this. By definition, `preferredDuringSchedulingIgnoredDuringExecution` is not a hard constraint, and would enable the scheduler to ignore these requirements if they were unavailable on the cluster and instead schedule on a node that has a non-working or degraded topology.

Additionally, this would suggest that the scope of the KEP could expand beyond just node labels, and instead define a set of well-known standard namespaces that could be applied in any suitable API data data model (e.g., node label, resourceslice [a DRA-provided resource], deviceclass [another DRA-provided resource], etc).

wdyt?

Reply:
You can always use `requiredDuringSchedulingIgnoredDuringExecution` for hard constraint.
I used `preferredDuringSchedulingIgnoredDuringExecution` as an example.